src/Data/ListLike/Base.hs

```diff
@@ -157,18 +156,16 @@ class (IsList full, item ~ Item full, FoldableLL full item, Monoid full) =>
 
     {- | Reverse the elements in a list. -}
     reverse :: full -> full
-    reverse l = rev l empty
-        where rev rl a
-                | null rl = a
-                | otherwise = rev (tail rl) (cons (head rl) a)
+    reverse = foldl' (\ a x -> cons x a) empty
+
     {- | Add an item between each element in the structure -}
     intersperse :: item -> full -> full
-    intersperse sep l
-        | null l = empty
-        | null xs = singleton x
-        | otherwise = cons x (cons sep (intersperse sep xs))
-        where x = head l
-              xs = tail l
+    intersperse sep l =
+        case uncons l of
+            Nothing -> empty
+            Just (x, xs)
+                | null xs -> singleton x
```

Comment:
See my comment on `init`: This extra lookahead should be avoidable by separating out a version for non-empty lists.